### PR TITLE
updating manager to support new Balancer rewarder

### PR DESCRIPTION
**Pull request by bulbozaur**

### CODE REVIEW

File: contracts/BalancerLiquidityGaugeWrapper.vy [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    log MinimalRewardsAmountUpdated(_new_min_rewards_amount)
+
+
+@external
+def set_weekly_amount(_new_weekly_amount: uint256):
+    """
+    @notice Sets the weekly amount. Can only be called by the owner.
+    """
+    assert msg.sender == self.owner, "not permitted"
+    self.weekly_amount = _new_weekly_amount
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply contracts/BalancerLiquidityGaugeWrapper.vy:256]: zero value can be used to stop contract

File: contracts/BalancerLiquidityGaugeWrapper.vy [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    log DistributorUpdated(_new_disributor)
+
+
+@external
+def set_min_rewards_amount(_new_min_rewards_amount: uint256):
+    """
+    @notice Sets the minimal LDO amount to start new period. Can only be called by the owner.
+    """
+    assert msg.sender == self.owner, "not permitted"
+    self.min_rewards_amount = _new_min_rewards_amount
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply contracts/BalancerLiquidityGaugeWrapper.vy:245]: zero value won't cause any issues

File: contracts/BalancerLiquidityGaugeWrapper.vy (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/BalancerLiquidityGaugeWrapper.vy:256]: maybe it's better to document this behavior at least :thinking:

File: contracts/BalancerLiquidityGaugeWrapper.vy (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/BalancerLiquidityGaugeWrapper.vy:245]: ok, thx

File: README.md [hunk 3]
```diff
@@ -1,133 +1,65 @@
...
+`brownie test -s`
 
-Returns current allowance limit available for distribution by calling `createDistribution`
+## Poetry
 
-##### `createDistribution(_token: address, _merkle_root: bytes32, _amount: uint256, _distribution_id: uint256):`
+`poetry install`
 
-Wrapper for `createDistribution` of MerkleOrchard contract.
+`poetry console`
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply README.md]: fixed

File: contracts/RewardsManager.vy [hunk 4]
```diff
@@ -1,51 +1,173 @@
...
+
+event RewardsContractUpdated:
+    newRewardsContract: indexed(address)
+
+
+event RewardsContractTransferred:
+    newDistributor: indexed(address)
+
+
+event WeeklyRewardsAmountUpdated:
```
Comment (arwer13) [inline contracts/RewardsManager.vy:37]: There are no tests for events `NewRewardsPeriodStarted` and `RewardsContractTransferred`

File: scripts/deploy.py [hunk 5]
```diff
@@ -1,59 +1,38 @@
 import sys
-import time
-from brownie import (
-    BalancerRewardsController,
-    RewardsManager
-)
+from brownie import RewardsManager
```
Comment (arwer13) [inline scripts/deploy.py:2]: I'm not sure, do we need a test for deploy script?

**Review by arwer13 — CHANGES_REQUESTED** [review]
Single contract version is much more appealing for me 👍

File: contracts/RewardsManager.vy [hunk 6]
```diff
@@ -1,51 +1,173 @@
...
+        rewards_amount = self.weekly_amount
+    
+    if iteration < 3: 
+        self.rewards_iteration = iteration + 1
+    else: 
+        self.rewards_iteration = 0
+
+    assert rewards_amount > 0, "manager: rewards disabled"
+    assert amount >= rewards_amount, "manager: low balance"
+    assert self._is_balancer_rewards_period_finished(rewards_contract), "manager: rewards period not finished"
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply contracts/RewardsManager.vy:141]: rewards_amount can be changed before. period finished assert moved

File: contracts/RewardsManager.vy [hunk 7]
```diff
@@ -54,33 +176,58 @@ def is_rewards_period_finished() -> bool:
...
+    current_owner: address = self.owner
+    assert msg.sender == current_owner, "not permitted"
+    self.owner = _to
+    log OwnershipTransferred(current_owner, _to)
 
 
-@view
 @external
-def period_finish() -> uint256:
+def transfer_rewards_contract(_to: address):
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply contracts/RewardsManager.vy:194]: next setter updates rewards contract, but this one transfers distribution permission to smb

File: contracts/RewardsManager.vy (hunk 4 above)
Comment (bulbozaur) [reply contracts/RewardsManager.vy:37]: RewardsContractTransferred has one, added for NewRewardsPeriodStarted

File: contracts/RewardsManager.vy [hunk 8]
```diff
@@ -1,51 +1,173 @@
...
+        self.rewards_iteration = 0
+
+    assert rewards_amount > 0, "manager: rewards disabled"
+    assert amount >= rewards_amount, "manager: low balance"
+    assert self._is_balancer_rewards_period_finished(rewards_contract), "manager: rewards period not finished"
+
+    ERC20(LDO_TOKEN).approve(rewards_contract, rewards_amount)
+    BalancerLiquidityGauge(rewards_contract).deposit_reward_token(LDO_TOKEN, rewards_amount)
+
+    log NewRewardsPeriodStarted(rewards_amount)
```
[the remark this answers is not in the document]
Comment (bulbozaur) [reply contracts/RewardsManager.vy:146]: not clear for me, it fires weekly with new reward period

File: scripts/deploy.py (hunk 5 above)
Comment (bulbozaur) [reply scripts/deploy.py:2]: Usually, I do test for deploy script, but this one contains only contracts deploy call, so I've used contact deploy directly in test as it'll be used on deployment

File: README.md [hunk 9]
```diff
@@ -1,126 +1,99 @@
...
+#### RewardsManager.sol
 
-##### `set_balancer_distributor(_new_balancer_distributor: address)`
+**def period_finish() -> uint256: view**
 
-Changes `BALANCER_DISTRIBUTOR`. Can be called by owner or current balancer_distributor.
+Returns estimated date of last rewards period start date
 
-Events:
+    BLG.periodFinish + (LDO.balanceOf(self)/self.rewardAmount) * SECONDS_PER_WEEK
```
Comment (TheDZhon) [inline README.md:36]: The formula differs from the implementation provided here: https://github.com/lidofinance/balancer-rewards-manager/blob/feature/v3/contracts/RewardsManager.vy#L148:L159

Also, the doc comment also has a slightly different meaning.

File: contracts/RewardsManager.vy [hunk 10]
```diff
@@ -1,51 +1,171 @@
...
-    self.rewards_contract = _rewards_contract
+    return self._balancer_period_finish(self.rewards_contract)
+
+
+@external
+def start_next_rewards_period():
+    """
+    @notice
+        Starts the next rewards period of duration `rewards_contract.deposit_reward_token(address, uint256)`,
+        distributing `self.weekly_amount` tokens throughout the period. The current
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:111]: Maybe throughout each week of the period?

File: contracts/RewardsManager.vy [hunk 11]
```diff
@@ -1,51 +1,171 @@
...
+    rewards_contract: address = self.rewards_contract
+    amount: uint256 = ERC20(LDO_TOKEN).balanceOf(self)
+    iteration: uint256 = self.rewards_iteration    
+    rewards_amount: uint256 = 0
+
+    assert rewards_contract != ZERO_ADDRESS, "manager: rewards disabled"
+    assert self._is_balancer_rewards_period_finished(rewards_contract), "manager: rewards period not finished"
+
+    if iteration == 0:
+        amount_to_distribute: uint256 = ERC20(LDO_TOKEN).balanceOf(self)
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:125]: Unused var

File: README.md [hunk 12]
```diff
@@ -1,126 +1,99 @@
...
+    amount: uint256
 ```
 
+```vyper=
+event WeeklyRewardsAmountUpdated:
+    newWeeklyRewardsAmount: uint256
+```
 
-##### `set_balancer_distributor(_new_balancer_distributor: address)`
+**def balancer_period_finish() -> uint256:**
```
Comment (TheDZhon) [inline README.md:58]: `view` keyword is omitted here

File: README.md [hunk 13]
```diff
@@ -1,126 +1,99 @@
...
 
-Changes `BALANCER_DISTRIBUTOR`. Can be called by owner or current balancer_distributor.
+Returns timestamp of current period ending at Balancer Liquidity Gauge
 
-Events:
+**def is_balancer_rewards_period_finished() -> bool:**
```
Comment (TheDZhon) [inline README.md:62]: `view` keyword is omitted

File: contracts/RewardsManager.vy [hunk 14]
```diff
@@ -54,33 +174,58 @@ def is_rewards_period_finished() -> bool:
...
+    assert msg.sender == self.owner, "not permitted"
+    assert _to != ZERO_ADDRESS, "zero address not allowed"
+    BalancerLiquidityGauge(self.rewards_contract).set_reward_distributor(LDO_TOKEN, _to)
+
+    log RewardsContractTransferred(_to)
+
 
 @external
-def start_next_rewards_period():
+def set_rewards_contract(_rewards_contract: address):
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:204]: Do we need to ensure that there is no ongoing reward distribution period before changing the contract?

File: contracts/RewardsManager.vy [hunk 15]
```diff
@@ -92,4 +237,5 @@ def recover_erc20(_token: address, _amount: uint256, _recipient: address = msg.s
     """
     assert msg.sender == self.owner, "not permitted"
     if _amount != 0:
-        assert ERC20(_token).transfer(_recipient, _amount), "token transfer failed"
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:95]: Let's add a check that `_recipient` is not zero.

File: contracts/RewardsManager.vy [hunk 16]
```diff
@@ -1,51 +1,171 @@
...
+event ERC20Recovered:
+    token: indexed(address)
+    amount: uint256
+    recipient: indexed(address)
 
 
 owner: public(address)
 rewards_contract: public(address)
-ldo_token: constant(address) = 0x5A98FcBEA516Cf06857215779Fd812CA3beF1B32
+min_rewards_amount: public(uint256)
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:53]: There is no additional setter for the `min_rewards_amount` field. Thus, may use `immutable` or add setter. What is better?

File: contracts/RewardsManager.vy [hunk 17]
```diff
@@ -1,51 +1,171 @@
...
+rewards_iteration: public(uint256)
+LDO_TOKEN: constant(address) = 0x5A98FcBEA516Cf06857215779Fd812CA3beF1B32
+SECONDS_PER_WEEK: constant(uint256) = 7 * 24 * 60 * 60
+WEEKS_PER_PERIOD: constant(uint256) = 4
 
 
 @external
-def __init__():
-    self.owner = msg.sender
+def __init__(
```
Comment (TheDZhon) [inline contracts/RewardsManager.vy:62]: Who is the initial reward distributor? Do we need to set him up during the construction stage?

File: tests/test_recover.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+from utils.config import (
+    lido_dao_agent_address,
+    ldo_token_address,
+    lido_dao_voting_address,
+    lido_dao_token_manager_address
+)
+from utils.evm_script import encode_call_script
+from utils.voting import create_vote
+
+
```
Comment (TheDZhon) [inline tests/test_recover.py:12]: I propose to add a test for a non-conformant ERC-20 token (USDT, for example). This one would leverage our low-level `safeTransfer` call.

**Review by TheDZhon — CHANGES_REQUESTED** [review]
LGTM except for some insignificant changes. :+1:

File: contracts/RewardsManager.vy (hunk 14 above)
Comment (bulbozaur) [reply contracts/RewardsManager.vy:203]: I think we don't, starting new period after previous finished is limitation of permissionless logic of current contract, we should be able to transfer distribution permission anyway if that will be required

**Review by TheDZhon — APPROVED** [review]
LGTM :rocket:

**Review by arwer13 — APPROVED** [review]
Good to go! 🕺

**Review by crazymonkey21y — APPROVED** [review]

**Comment by crazymonkey21y** [comment]
> _ No description provided. _

I didn't approve this

**Comment by crazymonkey21y** [comment]
The merger was a mistake I was doing this all for my phone and I didn't agree to the mergers I don't agree to the mergers

**Comment by crazymonkey21y** [comment]
The merger was a mistake I was doing this all for my phone and I didn't agree to the mergers I don't agree to the n*****

**Pull request merged**